phoenix-queryserver/src/main/java/org/apache/phoenix/queryserver/server/customizers/prometheus/PrometheusHadoopServlet.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.phoenix.queryserver.server.customizers.prometheus;
+
+import java.io.IOException;
+import java.io.Writer;
+import java.util.Collection;
+import java.util.regex.Pattern;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.apache.hadoop.metrics2.AbstractMetric;
+import org.apache.hadoop.metrics2.MetricType;
+import org.apache.hadoop.metrics2.MetricsRecord;
+import org.apache.hadoop.metrics2.MetricsTag;
+import org.apache.hadoop.metrics2.impl.MetricsExportHelper;
+
+public class PrometheusHadoopServlet extends HttpServlet {
+    private static final Pattern SPLIT_PATTERN =
+            Pattern.compile("(?<=[a-z])(?=[A-Z])|(?<=[A-Z])(?=([A-Z][a-z]))|\\W|(_)+");
+
+    @Override
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws IOException {
+        writeMetrics(resp.getWriter(), "true".equals(req.getParameter("description")),
+                req.getParameter("qry"));
+    }
+
+    static String toPrometheusName(String metricRecordName, String metricName) {
+        String baseName = metricRecordName + metricName.substring(0,1).toUpperCase() + metricName.substring(1);
+        String[] parts = SPLIT_PATTERN.split(baseName);
+        return String.join("_", parts).toLowerCase();
+    }
+
+    void writeMetrics(Writer writer, boolean descriptionEnabled, String queryParam)
+            throws IOException {
+        Collection<MetricsRecord> metricRecords = MetricsExportHelper.export();
+        for (MetricsRecord metricsRecord : metricRecords) {
+            for (AbstractMetric metrics : metricsRecord.metrics()) {
+                if (metrics.type() == MetricType.COUNTER || metrics.type() == MetricType.GAUGE) {
+
+                    String key = toPrometheusName(metricsRecord.name(), metrics.name());
```

Comment:
Note that this method and its regexp could be called quite often: per metricRecrod, per metric, per metric scrape interval ([default scrape interval is 1 minute](https://github.com/prometheus/prometheus/blob/6fc5305ce904a75f56ca762281c7a1b052f19092/config/config.go#L142)) could add to a bit of CPU time... if I am right, I suggest [memoizing this function](https://dzone.com/articles/java-8-automatic-memoization)